src/node/history.h

```diff
@@ -731,7 +731,20 @@ namespace ccf
       if (store.commit_gap() >= sig_tx_interval)
       {
         mguard.unlock();
-        emit_signature();
+
+        // Note: emit signature asynchronously to avoid possible deadlocks in
+        // case the signature synchronously triggers commit hooks that acquire
+        // already-taken locks (typically with the /node/create transaction
+        // triggered from the node state).
+        auto emit_sig_msg = std::make_unique<threading::Tmsg<EmitSigMsg>>(
```

Comment:
Because of the current fairness strategy between incoming ring buffer messages and thread messages, this currently means that in the very worst case, a signature may be scheduled 256 `seqno`s after the actual `tx_count` signature interval.

https://github.com/microsoft/CCF/blob/84461e142c56a6d969faefaec15920dc95867036/src/enclave/enclave.h#L414

There's a way to schedule the signature at its exact interval by modifying the fairness logic. I'll push the change in a separate commit (but same PR) so we can have a look at it separately.

Reply:
See https://github.com/microsoft/CCF/pull/3782/commits/54cfc533ec14fb31810c1d0d40795401cf646612